ogc/servers.py

```diff
@@ -103,25 +104,32 @@ def method():
             )  # bind route function call to instance method
 
     def ogc_render(self, ogc_idx):
-        print("OGC server.ogc_render", ogc_idx)
+        logger.info("OGC server.ogc_render %i", ogc_idx)
+        if request.method != "GET":
+            return respond_xml("<p>Only GET supported</p>", status=405)
+
         ogc = self.ogcs[ogc_idx]
         if not request.args:
             return self.home_func(ogc.endpoint)
         try:
+            # We'll filter out any characters from URl parameter values that are not in the allowlist.
+            # Note the parameter with key "params" has a serialized JSON value, leading to {}"
+            match_unapproved_chars = r'[^-A-Za-z0-9+.,_/:\{\}"]'
```

Comment:
regular expressions are so hard. What's in the unapproved_chars list? certainly A-Z and a-z and 0-9 are allowed? 

Checking some expressions here https://pythex.org/ it seems like it allows something like this `java:{test},_{}` but not this `<>&%$#@!*()`. So I think the code is correct, but the variable name `unapproved_chars` should maybe be `approved_chars`?

Reply:
The leading `^` is a not, so this is indeed a list of unapproved characters. Do we feel that this check is good enough, or should we be using an external library to extend our checking capabilities?  I don't want to add external libraries for no reason.

Reply:
Sorry, i've gotten so accustomed to regexes that I forget they can be a bit opaque.  Below is another form I'd considered, with some extra commentary - does that make it easier to read?

````python
allowed_chars = r'-A-Za-z0-9+.,_/:\{\}"' # Allowed chars are -, A through Z, a through z, 0 through 9, and the characters + . , _ / : { }
match_one_unallowed_char = "[^%s]"
re.sub(match_one_unallowed_char, "", str(v)) # Find every unallowed char and replace it with nothing (remove it)
````

Reply:
I thought `^` matched the start of the line, but there you go. Learn something new every day. 

@jdw-creare the new version is MUCH easier to read (with the note that `^` is the `not` operator). 

Longer answers. I believe our server enables UDP's, which send full PODPAC pipeline definitions through the URL... though I also believe that string is base64 encoded before reaching our server -- I'm not sure at what point it gets decoded and where in the pipeline this sanitization is happening. If it's the decoded, non-base64 encoded string, then to enable UDPs we also need all the arithmetic operators `*/+-`, and these `[]()` .

Reply:
Yeah it looks like the term for `[^...]` may be "negated set" - that's what they call it on this regex explainer: regexr.com/7tdg4  But you're remembering correctly - `^` at the start of a line matches the start of the input.

I'll push an update shortly!